compiler/surface/parser.mly

```diff
@@ -33,583 +33,649 @@ end>
    so that we concat them efficiently as much as possible. *)
 %right LAW_TEXT
 
+(* Precedence of expression constructions *)
+%right top_expr
+%right ALT
+%right let_expr IS
+%right logical_expr AND
+%right logical_or_expr OR XOR
+%nonassoc compare_expr GREATER GREATER_EQUAL LESSER LESSER_EQUAL EQUAL NOT_EQUAL
```

Comment:
Ah there is something here I wanted to ask here too: the usual convention is to have AND bind tighter than OR. Here we are using the opposite. Is there a specific reason, is that more natural related to natural language ?

Ideally compound OR/AND formula should require parens, IMHO. Since we still want to accept `a AND b AND c`, we can't make them non-associative, so that's not something we can encode with priorities. It could be done with nested rules though:
```
compare_expr ::= ...
and_expr ::= compare_expr AND and_expr
or_expr ::= compare_expr OR or_expr
expr ::= ... | compare_expr | or_expr | and_expr
```
but I am not sure how the conflicts here could be handled. That's also what I meant when saying that in some cases using priorities is less expressive — which may not be a bad thing.

Lastly, we could put both `AND` and `OR` at the same priority level, with associativity, meaning that `a OR b AND c OR d` would be `a OR (b AND (c OR d))` (or even `((a OR b) AND c) OR d`). I am not sure what is less error-prone between priorities or this actually…

Reply:
Fixed by #373